### PR TITLE
Key prefix iteration

### PR DESCRIPTION
**Pull request by cheme**
This PR add a method to iterator in order to iterate only on a prefix. This is a variant of 'seek' that will stop at the end of 'prefix' iteration ('technically we just trim the trail on 'prefix' call).

This is needed to avoid fetching a node after the prefix when doing a prefix iterator in substrate.
https://github.com/paritytech/substrate/blob/990acfe4d67c70b44005350a2e2122e810809849/primitives/state-machine/src/trie_backend_essence.rs#L158 is always fetching a node that is not needed and that will appear in call proof.
Replacing 'seek' by this new 'prefix' function (and removing the end condition) the proof will not have this additional node.

### CODE REVIEW

**Comment by jimpo** [comment]
I don't see why this needs to be part of `TrieIterator`. Can't this be implemented with a `TrieIterator` impl that wraps another iterator, seeks to the beginning of the prefix and stops when the next item is out of range?

**Comment by cheme** [comment]
Yes it seems like a better way of doing thing, have prefix function include into the iterator implementation. I will implement that later, putting in progress for now.

**Comment by cheme** [comment]
I did try it, but in fact I cannot change the trail (except if I expose the trail through trieiterator trait but that would be worse), I could of course iterate up to prefix start, but the pr will be useless as it will require fetching an extra node to assert end of iteration.
So I do not really think of another way to implement that.

**Comment by cheme** [comment]
The point of the pr is to avoid fetching "the next item is out of range" so in substrate this next item is not included in the proof (given the proof recorder use for call which simply record all fetched keys).

**Comment by cheme** [comment]
Looking at #42 I think I see how it could be implemented with a struct over the iterator, still I am not sure about this move; it makes sense for me to have two different primitives as they do something different: one position a cursor and another restrict the scope of the iterator.
@jimpo, For switching substrate prefix iterator (broken in 0.15), I am asking myself if we should already use this iterator (we can still change later as it will only impact produced proof but not state)?

**Comment by cheme** [comment]
@jimpo I did remove 'prefix' from triedb trait. It seems to play nicely with current substrate codebase: https://github.com/paritytech/substrate/compare/master...cheme:trie_prefix_iter

File: trie-db/src/iterator.rs [hunk 1]
```diff
@@ -221,6 +228,38 @@ impl<'a, L: TrieLayout> TrieIterator<L> for TrieDBNodeIterator<'a, L> {
...
 	}
+
+	/// Advance the iterator into a prefix, no value out of the prefix will be accessed
+	/// or returned after this operation.
+	pub fn prefix(&mut self, prefix: &[u8]) -> Result<(), TrieHash<L>, CError<L>> {
+		self.trail.clear();
+		self.key_nibbles.clear();
+		if self.seek_prefix(prefix)? {
+			if let Some(v) = self.trail.pop() {
+				if self.trail.len() > 0 {
```
Comment (jimpo) [inline trie-db/src/iterator.rs]: nit: This `if` check is unnecessary, you can just do

```rust
self.trail.clear();
self.trail.push(v);
```

which works in both cases.

File: trie-db/src/iterator.rs [hunk 2]
```diff
@@ -221,6 +228,38 @@ impl<'a, L: TrieLayout> TrieIterator<L> for TrieDBNodeIterator<'a, L> {
...
 		}
 	}
+
+	/// Advance the iterator into a prefix, no value out of the prefix will be accessed
+	/// or returned after this operation.
+	pub fn prefix(&mut self, prefix: &[u8]) -> Result<(), TrieHash<L>, CError<L>> {
+		self.trail.clear();
+		self.key_nibbles.clear();
+		if self.seek_prefix(prefix)? {
+			if let Some(v) = self.trail.pop() {
```
Comment (jimpo) [inline trie-db/src/iterator.rs]: Clever.

File: trie-db/src/iterator.rs [hunk 3]
```diff
@@ -221,6 +228,38 @@ impl<'a, L: TrieLayout> TrieIterator<L> for TrieDBNodeIterator<'a, L> {
 			node_hash = next_node_hash;
 		}
 	}
+
+	/// Advance the iterator into a prefix, no value out of the prefix will be accessed
+	/// or returned after this operation.
+	pub fn prefix(&mut self, prefix: &[u8]) -> Result<(), TrieHash<L>, CError<L>> {
+		self.trail.clear();
```
Comment (jimpo) [inline trie-db/src/iterator.rs]: These two things are already cleared in `seek_prefix`.

File: trie-db/src/iterator.rs [hunk 4]
```diff
@@ -95,8 +95,11 @@ impl<'a, L: TrieLayout> TrieDBNodeIterator<'a, L> {
 	}
 }
 
-impl<'a, L: TrieLayout> TrieIterator<L> for TrieDBNodeIterator<'a, L> {
-	fn seek(&mut self, key: &[u8]) -> Result<(), TrieHash<L>, CError<L>> {
+impl<'a, L: TrieLayout> TrieDBNodeIterator<'a, L> {
+	fn seek_prefix(
```
Comment (jimpo) [inline trie-db/src/iterator.rs]: Please leave a method doc explaining what the return value means.

**Comment by ordian** [comment]
@cheme please resolve the merge conflicts

**Comment by cheme** [comment]
Resolved the conflict, and also fix dependencies and import of fuzzing (was broken).

**Pull request merged**